src/types.rs

```diff
@@ -32,8 +42,227 @@ impl EdgeKind {
+pub(crate) fn least_upper_bound(
+    mut tags: impl Iterator<Item = Option<TypeBound>>,
+) -> Option<TypeBound> {
+    tags.fold_while(Some(TypeBound::Eq), |acc, new| {
+        if let (Some(acc), Some(new)) = (acc, new) {
+            Continue(Some(acc.union(new)))
+        } else {
+            // if any type is unbounded, short-circuit
+            Done(None)
+        }
+    })
+    .into_inner()
+}
+
+#[derive(Clone, PartialEq, Debug, Eq, derive_more::Display)]
+/// Core types: primitive (leaf), tuple (product) or sum (co-product).
+enum TypeEnum {
+    Prim(PrimType),
+    #[display(fmt = "Tuple({})", "_0")]
+    Tuple(TypeRow),
+    #[display(fmt = "Sum({})", "_0")]
+    Sum(TypeRow),
+}
+impl TypeEnum {
+    /// The smallest type bound - if any - that covers the whole type.
+    fn least_upper_bound(&self) -> Option<TypeBound> {
+        match self {
+            TypeEnum::Prim(p) => p.bound(),
+            TypeEnum::Tuple(ts) => least_upper_bound(ts.iter().map(Type::least_upper_bound)),
+            TypeEnum::Sum(ts) => least_upper_bound(ts.iter().map(Type::least_upper_bound)),
+        }
+    }
+}
+
+#[derive(
+    Clone, PartialEq, Debug, Eq, derive_more::Display, serde::Serialize, serde::Deserialize,
+)]
+#[display(fmt = "{}", "_0")]
+#[serde(into = "serialize::SerSimpleType", from = "serialize::SerSimpleType")]
+/// A HUGR type - the valid types of [EdgeKind::Value] and [EdgeKind::Static] edges.
+/// Such an edge is valid if the ports on either end agree on the [Type].
+/// Types have an optional [TypeBound] which places limits on the valid
+/// operations on a type.
+///
+/// Examples:
+/// ```
+/// # use hugr::types::{Type, TypeBound};
+/// # use hugr::type_row;
+///
+/// const unit: Type = Type::new_unit();
+/// let sum = Type::new_sum(type_row![unit, unit]);
+/// assert_eq!(sum.least_upper_bound(), Some(TypeBound::Eq));
+///
+/// ```
+///
+/// ```
+/// # use hugr::types::{Type, TypeBound, AbstractSignature};
+///
+/// let graph_type = Type::new_graph(AbstractSignature::new_linear(vec![]));
+/// assert_eq!(graph_type.least_upper_bound(), Some(TypeBound::Copyable));
+///
+/// ```
+///
+pub struct Type(TypeEnum, Option<TypeBound>);
+
+impl Type {
+    /// Initialize a new graph type with a signature.
+    pub fn new_graph(signature: AbstractSignature) -> Self {
+        Self::new(TypeEnum::Prim(PrimType::Graph(Box::new(signature))))
+    }
+
+    /// Initialize a new tuple type by providing the elements..
+    #[inline(always)]
+    pub fn new_tuple(types: impl Into<TypeRow>) -> Self {
+        Self::new(TypeEnum::Tuple(types.into()))
+    }
+
+    /// Initialize a new sum type by providing the possible variant types.
+    #[inline(always)]
+    pub fn new_sum(types: impl Into<TypeRow>) -> Self {
+        Self::new(TypeEnum::Sum(types.into()))
+    }
+
+    /// Initialize a new custom type.
+    // TODO remove? Resources/TypeDefs should just provide `Type` directly
+    pub const fn new_extension(opaque: CustomType) -> Self {
```

Comment:
Would `new_opaque` be a better name?

Reply:
I'm not sure I like referring to things as opaque any more - quite often they
are not actually opaque, just not core.

Reply:
not that I like `new_extension` much either